libc/include/llvm-libc-types/float16.h

```diff
@@ -0,0 +1,38 @@
+//===-- Definition of float16 type ----------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_TYPES_FLOAT16_H
+#define LLVM_LIBC_TYPES_FLOAT16_H
+
+#include "src/__support/macros/properties/architectures.h"
+#include "src/__support/macros/properties/compiler.h"
+#include "src/__support/macros/properties/cpu_features.h"
+
+#if defined(LIBC_TARGET_ARCH_IS_X86_64) && defined(LIBC_TARGET_CPU_HAS_SSE2)
+#if (defined(LIBC_COMPILER_CLANG_VER) && (LIBC_COMPILER_CLANG_VER >= 1500)) || \
+    (defined(LIBC_COMPILER_GCC_VER) && (LIBC_COMPILER_GCC_VER >= 1201))
+#define LIBC_TYPES_HAS_FLOAT16
+using float16 = _Float16;
+#endif
+#endif
+#if defined(LIBC_TARGET_ARCH_IS_AARCH64)
+#if (defined(LIBC_COMPILER_CLANG_VER) && (LIBC_COMPILER_CLANG_VER >= 900)) ||  \
+    (defined(LIBC_COMPILER_GCC_VER) && (LIBC_COMPILER_GCC_VER >= 1301))
+#define LIBC_TYPES_HAS_FLOAT16
+using float16 = _Float16;
+#endif
+#endif
+#if defined(LIBC_TARGET_ARCH_IS_ANY_RISCV)
+#if (defined(LIBC_COMPILER_CLANG_VER) && (LIBC_COMPILER_CLANG_VER >= 1300)) || \
+    (defined(LIBC_COMPILER_GCC_VER) && (LIBC_COMPILER_GCC_VER >= 1301))
+#define LIBC_TYPES_HAS_FLOAT16
+using float16 = _Float16;
+#endif
+#endif
```

Comment:
Moved from libc/src/__support/macros/properties/types.h.

We talked about just checking for `__FLT16_MANT_DIG__` but GCC < 13.1 defines it without supporting `_Float16` literals: https://godbolt.org/z/Pn9Ph6WPG.

Reply:
GCC 12 does support `_Float16` and it's the only one that does not support `f16` literal.  This can be checked with `defined(__FLT16_IS_IEC_60559__) && (__FLT16_IS_IEC_60559__ == 2)`.

Reply:
GCC >= 12.2 and < 13.1 also defines `__FLT16_MANT_DIG__` without supporting `_Float16` at all when compiling for AArch64. See https://godbolt.org/z/1d9E39fod.

Maybe we should just check for GCC 13.1 directly since `__FLT16_MANT_DIG__` is reliable on Clang:

```cpp
#if defined(__FLT16_MANT_DIG__) && (!defined(LIBC_COMPILER_IS_GCC) || LIBC_COMPILER_GCC_VER >= 1301)
#define LIBC_TYPES_HAS_FLOAT16
using float16 = _Float16;
#endif
```

Reply:
SGTM.